File: pe/nonvisualstudio/clojure/Implementation/Problem0001.clj
(ns pe.clojure.implementation.problem0001)

(require '[clojure.core.reducers :as r])

(defn isMultiple [number multipleOf]
  (= 0 (mod number multipleOf)))

(defn isMultipleOf3Or5 [number]
  (cond
   (isMultiple number 3) true
   (isMultiple number 5) true
   :else false))

(defn generateList [zeroUpToWhatNumber]
  (range 1 zeroUpToWhatNumber))

(defn sumOfMultiplesOfFiveAndThree [zeroUpToWhatNumber]
  (reduce
   +
   (filter
    isMultipleOf3Or5
    (generateList zeroUpToWhatNumber))))


;; Parallel - using reducers. pmap was a disaster. The problem size for filtering is too small for pmap.
;; This is a disaster as well.

(defn createVectorFromLazyList [listOfThings]
  (into
   []
   listOfThings))

(defn sumOfMultiplesOfFiveAndThreeWithReducer [zeroUpToWhatNumber]
  (reduce
   +
   (r/filter
    isMultipleOf3Or5
    (createVectorFromLazyList
     (generateList zeroUpToWhatNumber)))))

File: pe/nonvisualstudio/haskell/Implementation/Problem0001.hs
module Problem0001 where
	isMultiple number multipleOf = 0 == number `mod` multipleOf

	isMultipleOf3Or5 number | isMultiple number 3 	= True
							| isMultiple number 5 	= True
							| otherwise             = False

	sumOfMultiplesOfThreeAndFive maxNumberOfListExclusive =
		sum [x | x <- [1..(maxNumberOfListExclusive - 1)], isMultipleOf3Or5 x]
